src/core/net/checksum.cpp

```diff
@@ -115,6 +116,34 @@ void Checksum::Calculate(const Ip6::Address &aSource,
     }
 }
 
+void Checksum::Calculate(const Ip4::Address &aSource,
+                         const Ip4::Address &aDestination,
+                         uint8_t             aIpProto,
+                         const Message &     aMessage)
+{
+    Message::Chunk chunk;
+    uint16_t       length = aMessage.GetLength() - aMessage.GetOffset();
+
+    // Pseudo-header for checksum calculation (RFC-2460).
+    if (aIpProto != Ip4::kProtoIcmp)
+    {
+        AddData(aSource.GetBytes(), sizeof(Ip4::Address));
+        AddData(aDestination.GetBytes(), sizeof(Ip4::Address));
+        AddUint16(length);
+        AddUint16(static_cast<uint16_t>(aIpProto));
```

Comment:
Is this psudo header format valid for IPv4 and  ICMP? 
e.g., do we need to add the IpProto as `uint16`?

Reply:
This is validated by the unit tests added, added a comment around this.

Reply:
Thanks. 
Should we reorder these to match the definition of pseudo header for TCP/UDP in IPv4?

From https://datatracker.ietf.org/doc/html/rfc768  the `length` is at the end an after the `zero | protocol` field.
```
The pseudo  header  conceptually prefixed to the UDP header contains the
source  address,  the destination  address,  the protocol,  and the  UDP
length.   This information gives protection against misrouted datagrams.
This checksum procedure is the same as is used in TCP.

                  0      7 8     15 16    23 24    31
                 +--------+--------+--------+--------+
                 |          source address           |
                 +--------+--------+--------+--------+
                 |        destination address        |
                 +--------+--------+--------+--------+
                 |  zero  |protocol|   UDP length    |
                 +--------+--------+--------+--------+

```

It may not matter in checksum calculation but may be better to follow the order as defined by RFC.

BTW, we can consider removing the comment `// By converting aIpProto to uint16_t, we are actually appending eight bits of 0 before when calculating the presudo header.`. I think that's clear from the code. My question was to check the psudoe-header format (for TCP/UDP checksum calculation under IPv4) :)

Reply:
done